peer.go

```diff
@@ -0,0 +1,441 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gubernator
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/mailgun/errors"
+	"github.com/mailgun/holster/v4/clock"
+	"github.com/mailgun/holster/v4/collections"
+	"github.com/mailgun/holster/v4/ctxutil"
+	"github.com/mailgun/holster/v4/setter"
+	"github.com/mailgun/holster/v4/tracing"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type Peer struct {
+	lastErrs   *collections.LRUCache
+	wg         sync.WaitGroup
+	queue      chan *request
+	mutex      sync.RWMutex
+	client     PeerClient
+	conf       PeerConfig
+	inShutdown int64
+}
+
+type response struct {
+	rl  *RateLimitResponse
+	err error
+}
+
+type request struct {
+	request *RateLimitRequest
+	resp    chan *response
+	ctx     context.Context
+}
+
+type PeerConfig struct {
+	PeerClient PeerClient
+	Behavior   BehaviorConfig
+	Info       PeerInfo
+	Log        FieldLogger
+}
+
+type PeerClient interface {
+	Forward(context.Context, *ForwardRequest, *ForwardResponse) error
+	Update(context.Context, *UpdateRequest) error
+}
+
+func NewPeer(conf PeerConfig) (*Peer, error) {
+	if len(conf.Info.HTTPAddress) == 0 {
+		return nil, errors.New("Peer.Info.HTTPAddress is empty; must provide an address")
+	}
+
+	setter.SetDefault(&conf.PeerClient, NewPeerClient(WithNoTLS(conf.Info.HTTPAddress)))
+	setter.SetDefault(&conf.Log, logrus.WithField("category", "Peer"))
+
+	p := &Peer{
+		lastErrs: collections.NewLRUCache(100),
+		queue:    make(chan *request, 1000),
+		client:   conf.PeerClient,
+		conf:     conf,
+	}
+	go p.run()
+	return p, nil
+}
+
+// Info returns PeerInfo struct that describes this Peer
+func (p *Peer) Info() PeerInfo {
+	return p.conf.Info
+}
+
+var (
+	// TODO: Should retry in this case
+	ErrPeerShutdown = errors.New("peer is shutdown; try a different peer")
+)
+
+// Forward forwards a rate limit request to a peer.
+// If the rate limit has `behavior == BATCHING` configured, this method will attempt to batch the rate limits
+func (p *Peer) Forward(ctx context.Context, r *RateLimitRequest) (resp *RateLimitResponse, err error) {
+	ctx = tracing.StartNamedScope(ctx, "Peer.Forward")
+	defer func() { tracing.EndScope(ctx, err) }()
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(
+		attribute.String("peer.HTTPAddress", p.conf.Info.HTTPAddress),
```

Comment:
Pretty sure OTel/Honeycomb prefer dot.delimited.snake_case instead of PascalCase.

Suggest: `peer.http_address`